File: src/Striver_SDE_Sheet/Linked_List/Flattening_a_Linked_List/Flattening.cpp

Node *flatten(Node *root)
{
   
   vector<int> v;
   Node* temp=root;
   while(temp!=NULL){
       Node* b=temp;
       while(b!=NULL){
           v.push_back(b->data);
           b=b->bottom;
       }
       temp=temp->next;
   }
   sort(v.begin(),v.end());
   Node* head=new Node(0);
   Node* temp1=head;
   for(int i=0;i<v.size();i++){
        Node* ans=new Node(v[i]);
        temp1->bottom=ans;
        temp1=temp1->bottom;
   }
   temp1->next=NULL;
   return head->bottom;
}

File: src/Striver_SDE_Sheet/Linked_List/Flattening_a_Linked_List/Flattening.java
package Striver_SDE_Sheet.Linked_List.Palindrome_Linked_List;

public class Flattening {
    Node flatten(Node root)
    {
    	if(root == null || root.next ==  null)
    		return root;
    	
    	// Recursion for right lists
    	root.next = flatten(root.next);
    	
    	// Now merging them all
    	root = mergeTwoLists(root, root.next);
    	
    	return root; // Returning root in merge with its Left
    }
    
    Node mergeTwoLists(Node a, Node b){
    	// pointing variables
        Node temp = new Node(0); 
        Node res = temp;
        
        while(a != null && b != null){
        	// comparison and  Goes on
            if(a.data < b.data){ 
                temp.bottom = a;
                temp = temp.bottom;
                a = temp.bottom;
            }else{
                temp.bottom = b;
                temp = temp.bottom;
                b = temp.bottom;
            }
        }
        
        if(a != null) temp.bottom = a; 
        else temp.bottom = b;
        
        return res.bottom; // Head of flatten LL
    }
}